firebase-sessions/src/main/kotlin/com/google/firebase/sessions/api/FirebaseSessionsDependencies.kt

```diff
@@ -0,0 +1,93 @@
+ * limitations under the License.
+ */
+
+package com.google.firebase.sessions.api
+
+import android.util.Log
+import androidx.annotation.VisibleForTesting
+import java.util.Collections.synchronizedMap
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * [FirebaseSessionsDependencies] determines when a dependent SDK is installed in the app. The
+ * Sessions SDK uses this to figure out which dependencies to wait for to getting the data
+ * collection state. This is thread safe.
+ *
+ * This is important because the Sessions SDK starts up before dependent SDKs.
+ */
+object FirebaseSessionsDependencies {
+  private const val TAG = "SessionsDependencies"
+
+  private val dependencies = synchronizedMap(mutableMapOf<SessionSubscriber.Name, Dependency>())
+
+  /**
+   * Add a subscriber as a dependency to the Sessions SDK. Every dependency must register itself, or
+   * the Sessions SDK will never generate a session.
+   */
+  fun addDependency(subscriberName: SessionSubscriber.Name) {
+    if (dependencies.containsKey(subscriberName)) {
+      Log.d(TAG, "Dependency $subscriberName already added.")
+      return
+    }
+
+    // The dependency is locked until the subscriber registers itself.
+    dependencies[subscriberName] = Dependency(Mutex(locked = true))
+  }
+
+  /** Register and unlock the subscriber. This must be called before [getSubscribers] can return. */
+  internal fun register(subscriber: SessionSubscriber) {
+    val subscriberName = subscriber.sessionSubscriberName
+    val dependency = getDependency(subscriberName)
+
+    dependency.subscriber?.run {
+      Log.d(TAG, "Subscriber $subscriberName already registered.")
+      return
+    }
+    dependency.subscriber = subscriber
+
+    // Unlock to show the subscriber has been registered, it is possible to get it now.
+    dependency.mutex.unlock()
+  }
+
+  /** Gets the subscribers safely, blocks until all the subscribers are registered. */
+  internal suspend fun getSubscribers(): Map<SessionSubscriber.Name, SessionSubscriber> {
+    // The call to getSubscriber will never throw because the mutex guarantees it's been registered.
+    return dependencies.mapValues { (subscriberName, dependency) ->
+      dependency.mutex.withLock { getSubscriber(subscriberName) }
+    }
+  }
+
+  /** Gets the subscriber, regardless of being registered. This is exposed for testing. */
+  @VisibleForTesting
+  internal fun getSubscriber(subscriberName: SessionSubscriber.Name): SessionSubscriber {
+    return getDependency(subscriberName).subscriber
+      ?: throw IllegalStateException("Subscriber $subscriberName has not been registered.")
+  }
+
+  /** Resets all the dependencies for testing purposes. */
+  @VisibleForTesting internal fun reset() = dependencies.clear()
+
+  private fun getDependency(subscriberName: SessionSubscriber.Name): Dependency {
+    return dependencies.getOrElse(subscriberName) {
+      throw IllegalStateException(
```

Comment:
Sorry, a late nit pick: Should this throw an exception? Can we get away with a Log message? Basically when this exception occurs for a developers, there is nothing he would be able to fix it. So, thinking if this should be converted to a log message.

Reply:
I tried making this nullable, but it is pretty ugly to deal with nulls through this class. If the dependency hasn't been added then `getSubscriber` will end up throwing anyway, just with a less useful message. I intend this exception to help the 1P developer integrate a Firebase product with Sessions in the future.